### ISSUE TITLE
MFS disappeared after upgrade to 0.4.18

### ISSUE THREAD

**Issue by fiatjaf** [ipfs/go-ipfs#5738]
#### Version information:
go-ipfs version: 0.4.18-
Repo version: 7
System version: amd64/linux
Golang version: go1.11.1

#### Type:
Bug

#### Description:
Yesterday I was using 0.4.16, I was able to browse my MFS normally with `ipfs files ls /` and so on. Today `ipfs files ls /` shows nothing and `ipfs files ls /_` (previously a directory) shows `Error: file does not exist`.

**Comment by schomatis** [comment]
This does indeed seem related to the referenced issue since IPFS basically can't open the datastore, blocking until that one is fixed.

**Comment by Stebalien** [comment]
Note: It *also* looks like there was a bug in 0.4.16 that failed to flush MFS properly (ran into it when trying to test this). @schomatis is this something we fixed since then?

**Comment by schomatis** [comment]
Not that I recall (but 0.4.16 was 6 months ago and I wasn't that involved in MFS yet), could you post the commands to reproduce the flush bug?

**Comment by Stebalien** [comment]
1. Initialize a repo and start a daemon.
2. `ipfs files mkdir /a`
3. `ipfs files write --create /a/b <<<hello`
4. Kill the daemon with ^C
6. Start the daemon.
7. `ipfs files ls /a`.

**Comment by schomatis** [comment]
@Stebalien Running this in offline mode (without the daemon) does flush the changes, any idea what code might not be executed in one case or the other?

**Comment by Stebalien** [comment]
Are we debouncing flush?

**Comment by schomatis** [comment]
Couple of problems, the MFS Root `Close` is failing with `blockservice is closed` but no one is reporting this,

https://github.com/ipfs/go-ipfs/blob/254dda2823aaf9dca74deeff93e18b0dd1eb6b83/core/core.go#L683-L690

(speculation from this point on) there is a race a bit tricky to debug here (because if I a step into those instructions the `Close` works) but my best bet is in the republisher function which is run in a separate goroutine,

https://github.com/ipfs/go-mfs/blob/cb4fb10a6b6fb895ebd3bc46f7dff15dec288777/system.go#L253-L286

and the `publish` itself may be executed too late.

**Comment by schomatis** [comment]
@Stebalien WDYT?

**Comment by Stebalien** [comment]
Ah. So, I guess the *primary* issue here is that `Flush` isn't actually flushing. We should fix that but it might cause some other issues... Namely, we *don't* want to wait on *publishing*.

Really, this is where we want an event handler based system, not a republisher based system. That is, if we could attach event handlers that fire for *every* flush/update, we could use a rate-limited one for IPNS and a slightly throttled one for writing to disk.

**Comment by fiatjaf** [comment]
Don't know if this is relevant or not, but when I try to add a bunch of files (with a script) to my newly-created MFS directories, I see a bunch of

```
Error: file does not exist
Error: cannot store pin state: open /home/fiatjaf/.ipfs/datastore/033269.ldb: too many open files
```

But even with these errors happening for all of the files in the script, some of them actually are now being listed in the MFS directory, but not others.

When I try to add single files manually they seem to all get in there without errors being logged.

**Comment by schomatis** [comment]
> Really, this is where we want an event handler based system, not a republisher based system. That is, if we could attach event handlers that fire for _every_ flush/update, we could use a rate-limited one for IPNS and a slightly throttled one for writing to disk.

Is this something we can fix with the new MFS interface in https://github.com/ipfs/go-mfs/issues/3?

> When I try to add single files manually they seem to all get in there without errors being logged.

This seems more related to the `too many open files` errors than with MFS itself.

**Comment by Stebalien** [comment]
Ideally, yes.

**Comment by schomatis** [comment]
We can evaluate using an event handler based system in MFS (and there is a flush missing in the `ipfs files` command, there's a separate issue about this), but in the short term I'm not sure how to proceed here.

**Comment by fiatjaf** [comment]
I think this is tighly related to https://github.com/ipfs/go-ipfs/issues/5739 (because they're both happening in the same computer and repo). Although it's odd that I first noticed this and only later that, it would make sense to me if the second is causing the first.

**Comment by schomatis** [comment]
Yes, I agree. I'm closing this issue as we've opened separate ones to address the specific problems already mentioned here (that is: too many open files, event handler based system and missing flush in `ipfs files` command).
